Adding a Gradle Build system to the Processing IDE

Stefterv opened the pull request:
Here's the first pull-request adding a new Gradle build system to the Processing IDE.

The GitHub Actions will create distributable binaries for macOS (x64, aarch64), windows (x64) and linux (x64, aarch64). The other platforms will need their own CI/CD system, but I think the repo will already build on the other platforms.

In terms of compatibility I've come pretty far but I'm sure there is some functionality that I have missed. 

### New Features
- Installers for macOS (dmg), Widows (msi), linux (deb)
- Windows: Start menu item
- Jetpack Compose
- Kotlin Support for the IDE
- Splash screen Refactor + Animation
- Additional pde:// schema functions (To be documented)

### Functionality to test and verify
- [x] Examples
- [x] Tweaks mode
- [x] P2D
- [x] Debugger
- [X] Languages
- [X] Contributed libraries
- [X] Autocomplete
- [X] MovieMaker
- [X] Exporting Applications
- [x] `processing-java`
- [x] java mode default libraries (dxf,svg,io,net,serial,pdf) 
- [ ] Python mode (BROKEN, already on 4.3.1 macOS aarch64, other platforms okay)
- [x] Android mode

### To add during this pull request
- [x] file associations (windows ☑️, macOS ☑️, linux?)
- [x] pde:// prototcol (Probably works on windows and not macOS)
- [x] Webcam Usage (macOS)
- [x] Microphone Usage (macOS)
       https://github.com/JetBrains/compose-multiplatform/blob/master/tutorials/Native_distributions_and_local_execution/README.md#customizing-infoplist-on-macos
       https://github.com/JetBrains/compose-multiplatform/blob/master/tutorials/Signing_and_notarization_on_macOS/README.md#configuring-entitlements

### To improve during this pull request
- [x] Legacy resource handling (rn it copies all assets to temp folder on launch)
- [x] Move all dependencies to libs.versions.toml
- [x] Version and Build number

### After  pull request
- [ ] Signing (Windows, macOS, and Linux?)
        https://github.com/JetBrains/compose-multiplatform/blob/master/tutorials/Signing_and_notarization_on_macOS/README.md
- [ ] Releases CI/CD
- [x] linux (arm64) CI/CD
       This will require a custom runner setup as jetpack compose does not cross-build 
       We can setup the custom runner on Hertzner as they offer arm64 vps' for a low price
- [ ] Remove the raspberry pi runners as we will deprecate the arm32 build of Processing
- [ ] Update installation instructions on the website
- [ ] Installer customisation
- [ ] Flatpak distribution
### To consider
- Appstores

In any case, this pull request does not break the existing Ant build system so we can either wait with the merge or we can have the Gradle version available as a preview.

The bundle size has increased quite a bit since we are shipping a JRE and a JDK instead of using the JDK to run the IDE
|linux-x64 | 257 MB |  
|macos-aarch64 | 294 MB |  
|macos-x64 | 289 MB |  
|windows-x64 | 275 MB |

P.S. this also includes the pre-processor as a Maven library, so merging this PR will add processing:java:prepocessor as a Maven package

Fixes https://github.com/processing/processing4/issues/924

Stefterv committed:
run app with the jetpack compose runner

Stefterv committed:
Run app with Java Mode

Stefterv committed:
Deprecation indicators and removed unused resources

Stefterv committed:
Merge branch 'processing:main' into main-gradle

Stefterv committed:
Build sketches with new app bundler

Stefterv committed:
Added all modules for JAR filesystem

Stefterv committed:
Work with external JDK

Stefterv committed:
Distribution icons

Stefterv committed:
First functional Composable

Stefterv committed:
Proper styling

Stefterv committed:
Lottie animations

Stefterv committed:
Contribution Manager UI Test

Stefterv committed:
.idea meta files

Stefterv committed:
Show local contributions

Stefterv committed:
isUpdate & isInstalled & Contribution Pane

Stefterv committed:
Contribution panel

Stefterv committed:
Animation Test

Stefterv committed:
Added to the sketch menu

Stefterv committed:
Cleaned up ignored files

Stefterv committed:
Merge branch 'processing:main' into main-gradle

Stefterv committed:
backwards compatibility with ant

Stefterv committed:
re-add bin files

Stefterv committed:
.idea files

Stefterv committed:
gradle ci/cd

Stefterv committed:
Add generated code

Stefterv committed:
Update build-gradle.yml

Stefterv committed:
Create processing.png

Stefterv committed:
Update build-gradle.yml

Stefterv committed:
pre-processor gradle

Stefterv committed:
preprocessor decoupling

Stefterv committed:
Update build.gradle.kts

Stefterv committed:
windows fix

Stefterv committed:
LSP gradle build

Stefterv committed:
Added tests

Stefterv committed:
Fixed the tests

Stefterv committed:
Fixed regressions

Stefterv committed:
Update build.gradle.kts

Stefterv committed:
Fixed another regression

Stefterv committed:
Fixed another regression

Stefterv committed:
Fixed load keywords

Stefterv committed:
get installed JDK platform independent

Stefterv committed:
Reapply .gitignore

Stefterv committed:
fixed missing preproc output

Stefterv committed:
Update JavaBuild.java

Stefterv committed:
Embed JDK

Stefterv committed:
Set permissions in Gradle instead

Stefterv committed:
support windows

Stefterv committed:
Update build-gradle.yml

Stefterv committed:
Merge branch 'processing:main' into main-gradle

Stefterv committed:
Update build.yml

Stefterv committed:
Update pull_request.yml

Stefterv committed:
Cleanup pull request signing

Contributors do not have access to the repo's secrets

Stefterv committed:
Update pull_request.yml

Stefterv committed:
Update pull_request.yml

Stefterv committed:
Use shared resources

Found a better method that doesn't require copy the jar resources to a temporary folder

Stefterv committed:
JOGL Dependencies fix

The -main variants of JOGL trigger the natives to be included into the dependencies.
Appended to the copyCore task to also include the dependency jars

Stefterv committed:
Dependencies centralisation

Stefterv committed:
Dependencies centralisation 2

Stefterv committed:
Fixed missing modules & Windows meta

Stefterv committed:
Bugfix

Stefterv committed:
Jetpack Compose Compatibility

Calling Kotlin code while being backwards compatible with the Ant based build-system

Stefterv committed:
Download Examples

Download the processing from the GitHub and include it in the resources

Stefterv committed:
Java Mode resources and Libraries

Stefterv committed:
Fixed autocomplete and all other preprocessor issues

Stefterv committed:
Cleanup unused fix

Stefterv committed:
Run target

Stefterv committed:
Application export

Small macOS regression

Stefterv committed:
Fix install processing-java bug

Stefterv committed:
Bugfix

Stefterv committed:
Bugfix

Stefterv committed:
Removed different workfile

Stefterv committed:
Add macOS pde:// and file associations

Stefterv committed:
Camera and Microphone access

Stefterv committed:
Dynamic versioning

Stefterv committed:
Actions Versioning

Stefterv committed:
Fixed versioning error

Stefterv committed:
Update build-gradle.yml

Stefterv committed:
versioning

Stefterv committed:
bugfix

Stefterv committed:
bugfix

Stefterv committed:
Update build-gradle.yml

Stefterv commented:
`compose desktop` does not support Raspberry PI 32-bit, which would mean models older than 3B would become unsupported. The raspi 2 was released in 2015 and I'm not sure how well it would run Processing in any case, so could that be an argument to drop support for it?

Stefterv commented:
Another long term consideration I found is that `compose desktop` does not support building for another OS than the currently running one. For the IDE that is fine as we can use GitHub Actions. For exporting sketches this might be a bit limiting for users of Processing. I'd love to use the same packager within both contexts but we might have to use a different one for exporting sketches.

SableRaf commented:
> `compose desktop` does not support Raspberry PI 32-bit, which would mean models older than 3B would become unsupported. The raspi 2 was released in 2015 and I'm not sure how well it would run Processing in any case, so could that be an argument to drop support for it?

Sounds reasonable. We should communicate this early and gather community feedback. Could you draft an issue to explain the rationale for dropping 32-bit Raspberry Pi support? I'll share it on Discord, Discourse, and social media to request comments. Let me know if I can help in other ways.

SableRaf commented:
> I'd love to use the same packager within both contexts but we might have to use a different one for exporting sketches.

Thanks for flagging this. What would be the consequences of this approach? Larger download? Other limitations? Could we consider releasing platform-specific build support as tools/libraries? A bit like Unity does it?

Stefterv commented:
> Thanks for flagging this. What would be the consequences of this approach? Larger download? Other limitations? Could we consider releasing platform-specific build support as tools/libraries? A bit like Unity does it?

Consequences are not really clear for now as we're not really trying to do anything special while bundling. Which means we're also pretty flexible in terms of which bundler we use.

If we in the future would like to add composable UI's to Processing sketches for example, then we might need some of the functionality of the bundler that the IDE uses.

Stefterv committed:
New splash screen

Stefterv committed:
macOS tItle bar tweak

Stefterv committed:
Theme selector compatibility

Stefterv commented:
https://github.com/user-attachments/assets/e86d7611-243e-4f3a-99c1-1c0da0d059b3

Little preview of the new splash screen and the tweak to the macOS titlebar

Before

https://github.com/user-attachments/assets/580158b0-b359-462e-83db-c7bbb84ca812

Stefterv commented on java/test/processing/mode/java/ParserTests.java:
```diff
@@ -460,8 +460,9 @@ public void testMultilineStringClass() {
   public void testMultiMultilineString() {
-    Preferences.setBoolean("export.application.fullscreen", true);
-    expectGood("fullscreen_export");
+    // TODO: Add support for fullscreen. Not through settings. In PdeParseTreeListener.java
```
This can be re-enabled I think, it's been fixed

Stefterv commented on java/src/processing/mode/java/preproc/PreprocessIssueMessageSimplifier.java:
```diff
@@ -78,16 +74,8 @@ public static PreprocessIssueMessageSimplifier get() {
-      errStr = DefaultErrorLocalStrSet.get().get("editor.status.error.syntax").orElse("Error");
-      retStr = DefaultErrorLocalStrSet.get().get(stringName).orElse(stringName);
-    }
+    var errStr = DefaultErrorLocalStrSet.get().get("editor.status.error.syntax").orElse("Error");
```
Still not sure how to fix this

Stefterv commented on java/src/processing/mode/java/JavaBuild.java:
```diff
@@ -725,18 +732,20 @@ protected boolean exportApplication(File destFolder,
-                   new File(resourcesFolder, "en.lproj"));
-      Util.copyFile(mode.getContentFile("application/application.icns"),
-                    new File(resourcesFolder, "application.icns"));
+      // TODO: Handle the java embed and Icon with the new build system
```
TODO

Stefterv commented on app/src/processing/app/Platform.java:
```diff
@@ -382,8 +391,37 @@ static public File getContentFile(String name) {
+        // Default installation paths for different operating systems
+        if (os.contains("windows")) {
+          var programFiles = new File(System.getenv("ProgramFiles"));
+          return new File(programFiles, "Eclipse Adoptium/jdk-17.0.10+7-hotspot");
```
Not sure if hardcoded JDK-17 is still relevant

Stefterv committed:
Processing Website examples

The Processing Website has more meta-data related to the examples. I'd like to use that meta-data going forward so I've added it

Stefterv committed:
Windows look & feel

Stefterv commented:
https://github.com/user-attachments/assets/aadde617-dced-4bdf-9935-f6fb61f9736f

Windows adjusted titlebar

Stefterv committed:
Move contribution url into centralised place

Using the work by @mingness to simplify the contributions workflow.
Disabled the post request because that data is not saved anywhere and most hosts do not support it

Stefterv committed:
Move update check centralized

Stefterv committed:
New splash image

Stefterv committed:
Open new window if none are open

Stefterv committed:
Legacy Tasks explaination

Stefterv committed:
Improve debugging experience

Stefterv committed:
Documentation

Stefterv committed:
Examples for Core development

SableRaf committed:
Update BUILD.md

Fix typo/phrasing

Stefterv commented on app/src/processing/app/Mode.java:
```diff
@@ -501,7 +503,7 @@ public void rebuildImportMenu() {  //JMenu importMenu) {
 
     JMenuItem manageLibs = new JMenuItem(Language.text("menu.library.manage_libraries"));
-    manageLibs.addActionListener(e -> ContributionManager.openLibraries());
+    manageLibs.addActionListener(e -> openContributionsManager());
```
Revert back to original contribs manager

Stefterv commented on app/build.gradle.kts:
```diff
@@ -0,0 +1,315 @@
+    into(layout.buildDirectory.dir("resources-bundled/common/modes/java/examples"))
+}
+tasks.register<Download>("downloadProcessingWebsiteExamples") {
+    src("https://github.com/processing/processing-website/archive/refs/heads/main.zip")
```
This downloads .es files which break the examples

Stefterv committed:
Remove .es. and liveSketch from extended examples

Stefterv committed:
Merge branch 'main-gradle' of https://github.com/Stefterv/processing4 into main-gradle

Stefterv committed:
Re-enable fullscreen test

Stefterv committed:
Removed hardcoded JDK

Stefterv committed:
Back to original contributions manager

Stefterv committed:
Application export JDK and resources

SableRaf commented:
@Stefterv: Thanks for tagging me. What would you like me to review?

At first glance, the splash screen is still the old one. Tested on macOS 13.4.

This build also includes the experimental p5.js mode.

hx2A commented:
> At first glance, the splash screen is still the old one. Tested on macOS 13.4.

Side question: you have a macOS 13.4 machine? Does it have an Intel CPU or Apple Silicon?

SableRaf commented:
> Side question: you have a macOS 13.4 machine? Does it have an Intel CPU or Apple Silicon?

Yes I'm currently on 13.4 with an M1 (Apple Silicon)

Stefterv commented:
Testable versions can only be found on my fork here: https://github.com/Stefterv/processing4/actions/runs/12705347573

Stefterv committed:
fixed CI/CD

Stefterv committed:
Merge branch 'main' into main-gradle

Stefterv commented:
I've added the Gradle CI/CD to the pull requests. This will mean from now on we have an ant-based version and a gradle-based version building on pull requests. After the gradle version has been tested we can remove the ant based CI

Stefterv committed:
Ant compatibility

Stefterv committed:
Add Linux arm64 build step

Stefterv committed:
Only run on arm64 runners

Stefterv committed:
Only run on arm64 runners

Stefterv committed:
Merge branch 'processing:main' into main-gradle

Stefterv commented:
Windows arm64 compatibility seems to be in the current beta of `compose desktop` https://youtrack.jetbrains.com/issue/CMP-2584 so could be added as a support soon too

Requires Java 21!

Stefterv committed:
Merge remote-tracking branch 'upstream/main' into main-gradle

Stefterv committed:
Merge branch 'processing:main' into main-gradle

Stefterv committed:
More helpful artifact names

Stefterv committed:
Update pull_request.yml

Stefterv committed:
Issues 931 testing

Stefterv committed:
Merge branch 'processing:main' into main-gradle

SableRaf commented:
Fixes https://github.com/processing/processing4/issues/924

Stefterv committed:
Merge branch 'main' into main-gradle

Stefterv commented:
<img width="1191" alt="Screenshot 2025-01-25 at 10 23 51" src="https://github.com/user-attachments/assets/e3d9f0ce-8503-4fce-aa16-86ba0bfd163e" />
Found an issue on Raspberry PI OS 64-bit where the text do not display

Stefterv committed:
Fixed missing fonts on Linux

Stefterv commented:
<img width="1250" alt="Screenshot 2025-01-25 at 11 10 11" src="https://github.com/user-attachments/assets/7a4a12f9-33a1-4a1b-82c5-44648834935e" />
Fixed

Stefterv committed:
Remove dependency on Java mode and copy instead

Stefterv committed:
Better Kotlin-Ant compatibility

Stefterv committed:
New schema handler

Stefterv committed:
More Schema options

Stefterv committed:
Added support for Base64 encoded extra files

SableRaf commented:
The proposed visual update looks beautiful with the theme’s background color extending to the title bar. However, the contrast with the title text isn’t accessible in the current version. This will need adjustments to meet WCAG standards. 

<img width="688" alt="image" src="https://github.com/user-attachments/assets/4ef64a7e-83f8-4915-a019-d17c89d58c60" />

<img width="1150" alt="image" src="https://github.com/user-attachments/assets/5296b1b4-a069-459a-88d4-b2d76408f1e6" />

It's even worse when the window is out of focus:
<img width="693" alt="image" src="https://github.com/user-attachments/assets/48ffe722-87e9-4a4e-ad26-bcd02acad558" />

And some themes are worse than others:
<img width="692" alt="image" src="https://github.com/user-attachments/assets/e125d25d-7b11-4a6b-bab4-ef7429bae828" />

#### Edit: Windows 10

On Windows 10, the title looks fine when focused but out of focus, the title (and buttons) are not readable.

![image](https://github.com/user-attachments/assets/9d755e5b-0226-450c-bba7-4b7325f6f06d)

![image](https://github.com/user-attachments/assets/074b5f85-d44a-40b4-817e-97db3fbc5710)

SableRaf commented:
On Windows 10, switching theme doesn't immediately update the title bar color. A restart is necessary.

![image](https://github.com/user-attachments/assets/9f592301-3139-480c-b43c-5975d188d200)

Stefterv commented:
The contrast issue on macOS can only be solved by forcing the application in light mode, as the color of the text of the title bar is controlled by the OS
<img width="812" alt="Screenshot 2025-02-03 at 11 25 59" src="https://github.com/user-attachments/assets/df80bbf2-1ef4-490a-9668-c6eed350c876" />
I will look into the windows issue further, it can also be fixed by opening a new editor window. As far as I understand the title bar there is not changable after instatiantion. We can either re-open the editor after changing of the theme or wait until we switch that part over to Jetpack Compose

Stefterv committed:
- Imports cleanup
- Disable system look & feel on macOS for readability
- Cleaned unnecessary space
- Cleaned finished TODO

SableRaf commented:
> We can either re-open the editor after changing of the theme or wait until we switch that part over to Jetpack Compose

If that's not too much work, I'd say let's see how the re-opening of the editor feels. 

Alternatively, how about we inform the user that the full change will take effect after they restart Processing? Could be a pop up message.

Stefterv committed:
Removed outdated ant backwards compatibility marks

Stefterv committed:
Allow branches with `/` in the name

Stefterv committed:
Revert "Allow branches with `/` in the name"

This reverts commit 3b9aea1f40f1e7b903250cedaacc0e49f213ea9b.

Stefterv committed:
Gradle Release initial workflow setup

[skip ci]

Stefterv committed:
Merge branch 'processing:main' into main-gradle

Stefterv committed:
File associations

Stefterv committed:
Update build.gradle.kts

Stefterv committed:
Refactored / Simplified App Build script

Stefterv committed:
Github runners

Stefterv committed:
Update README.md

Stefterv committed:
Disable new schema for now

Stefterv committed:
Update build.gradle.kts

Stefterv committed:
Added jdk.accessibility

SableRaf reviewed (approved):
Thanks for your work on this PR @Stefterv! This is a huge leap forward for Processing 💙

hx2A commented:
Thank you, @Stefterv !! You worked very hard on this!
